src/App.test.js

```diff
@@ -1,8 +1,126 @@
-import { render, screen } from '@testing-library/react';
-import App from './App';
+import { render, screen, fireEvent } from "@testing-library/react";
+import App from "./App";
 
-test('renders learn react link', () => {
-  render(<App />);
-  const linkElement = screen.getByText(/learn react/i);
-  expect(linkElement).toBeInTheDocument();
+describe("App acceptance tests", () => {
+  describe("All existing user products are listed on the main page", () => {
+    beforeEach(() => {
+      render(<App />);
+    });
+
+    test("if in product listing page", () => {
+      const productListTitle = screen.getByText(/My wishlist/i);
+      expect(productListTitle).toBeInTheDocument();
+    });
+
+    test("if it renders the first product", () => {
+      const productItem = screen.getByText(/First product/i);
+      expect(productItem).toBeInTheDocument();
+    });
+
+    test("if it renders the second product", () => {
+      const productItem = screen.getByText(/Second product/i);
+      expect(productItem).toBeInTheDocument();
+    });
+
+    test("if it renders the third product", () => {
+      const productItem = screen.getByText(/Third product/i);
+      expect(productItem).toBeInTheDocument();
+    });
+  });
+
+  describe("It is possible to add a new product in the list", () => {
+    beforeEach(() => {
+      render(<App />);
+    });
+
+    describe("when the button to add a new product is clicked", () => {
+      beforeEach(() => {
+        const addNewProductButton = screen.getByTitle(/Add new product/i);
+        fireEvent.click(addNewProductButton);
+      });
+
+      test("if in edit product page", () => {
+        const editPageTitle = screen.getByText(/Product Details/i);
+        expect(editPageTitle).toBeInTheDocument();
+      });
+
+      describe("when typing product details and saving", () => {
+        const newProduct = {
+          name: "Test product name",
+        };
+
+        beforeEach(() => {
+          screen.debug();
+          // type required input fields
+          const productNameField = screen.getByPlaceholderText(/Product name .../i);
+          fireEvent.change(productNameField, {
+            target: { value: newProduct.name },
+          });
+          // click save button
+          const saveProductButton = screen.getByTitle(/Save product/i);
+          fireEvent.click(saveProductButton);
+        });
+
+        test("if returned to product listing page", () => {
+          const productListTitle = screen.getByText(/My wishlist/i);
+          expect(productListTitle).toBeInTheDocument();
+        });
+
+        test("if it renders the new product in product list", () => {
+          const productItem = screen.getAllByText(newProduct.name)[0];
+          expect(productItem).toBeInTheDocument();
+        });
+      });
+    });
+  });
+
+  describe("It is possible to edit an existing product in the list", () => {
```

Comment:
One of the points given to us is for "well documented tests". Huge kudos to including tests, each of use will have to cover our development with tests from now on, but they also have to include comments, I believe.

Reply:
I don't know what is their definition of "well documented", but as long as we have comments describing what's going on there - i hope that must be enough. Thank you, Bruno!